File: examples/energy/energy.py
hlp = """
    Test prediction with low-rank approximations on the energy dataset with matern, exponential and
    periodic kernels. Tested methods are Mklaren, ICD, CSI, Nystrom, FITC, RFF_KMP.

"""

if __name__ == "__main__":
    import matplotlib
    matplotlib.use("Agg")

import argparse
import csv
import os
import numpy as np
import itertools as it
from mklaren.kernel.kernel import exponential_kernel, kernel_sum, matern32_gpy, periodic_kernel
from mklaren.kernel.kinterface import Kinterface
from mklaren.regression.spgp import SPGP
from sklearn.metrics import mean_squared_error as mse
from datasets.energy import load_energy
from examples.inducing_points.inducing_points import plot_signal_subplots, test

# Hyperparameters
delta        = 10                                   # Look-ahead parameter
rank_range   = (7, )                                # Tested rank range
lambda_range = [0] + list(np.logspace(-1, 1, 5))    # L2-regularization parameter range


def process(dataset, kernel, outdir):
    """
    Run experiments with epcified parameters.
    :param dataset: Dataset key.
    :param kernel: Kernel function name.
    :param outdir: Output directory.
    :return:
    """

    kernel_function = None
    pars = dict()
    methods = []

    # Experiment parameters
    if kernel == "exponential":
        kernel_function = exponential_kernel
        pars = {"gamma": np.logspace(-4, 4, 5)}
        methods = ("Mklaren", "ICD", "CSI", "Nystrom", "SPGP", "RFF", "RFF-NS")

    elif kernel == "matern":
        kernel_function = matern32_gpy
        pars = {"lengthscale": SPGP.gamma2lengthscale(np.logspace(-4, 4, 5))}
        methods = ("Mklaren", "ICD", "CSI", "Nystrom", "SPGP")

    elif kernel == "periodic":
        kernel_function = periodic_kernel
        pars = {"sigma": np.logspace(-2, 2, 5),
                "per": np.logspace(1, 3, 10)}
        methods = ("Mklaren", "ICD", "CSI", "Nystrom", )


    # Data parameters
    signals = ["T%d" % i for i in range(1, 10)]
    assert dataset in signals
    inxs = range(1, 19)

    # Store results
    if not os.path.exists(outdir): os.makedirs(outdir)
    fname = os.path.join(outdir, "%s.csv" % dataset)
    subdname = os.path.join(outdir, "_%s" % dataset)
    if not os.path.exists(subdname):
        os.makedirs(subdname)

    header = ["experiment", "signal",  "tsi", "method",
              "mse_val", "mse_f", "mse_y", "rank", "lbd", "n", "time"]
    writer = csv.DictWriter(open(fname, "w", buffering=0), fieldnames=header)
    writer.writeheader()

    # Load energy consumption data
    data = load_energy(signal=dataset)
    data_out = load_energy(signal="T_out")
    X, labels = data["data"], data["labels"]
    X_out, _ = data_out["data"], data_out["labels"]

    # Normalize by outside temperature
    # Assume true signal is the mean of the signals
    Y = X - X_out
    N, n = Y.shape

    # Training, validation, test
    x  = np.atleast_2d(np.arange(0, n, 3)).T
    xv = np.atleast_2d(np.arange(1, n, 3)).T
    xp = np.atleast_2d(np.arange(2, n, 3)).T

    Nt, Nv, Np = x.shape[0], xv.shape[0], xp.shape[0]
    f = Y[1:19, x].mean(axis=0).ravel()

    for rank, lbd, tsi in it.product(rank_range, lambda_range, inxs):
        y = Y[tsi, x].reshape((Nt, 1))
        yv = Y[tsi, xv].reshape((Nv, 1))
        yp = Y[tsi, xp].reshape((Np, 1))

        # Sum and List of kernels
        vals = list(it.product(*pars.values()))
        names = pars.keys()
        Klist = [Kinterface(data=x, kernel=kernel_function, row_normalize=True,
                            kernel_args=dict([(n, v) for n, v in zip(names, vlist)])) for vlist in vals]
        Ksum = Kinterface(data=x, kernel=kernel_sum, row_normalize=True,
                            kernel_args={"kernels": [kernel_function] * len(vals),
                                         "kernels_args": [dict([(n, v) for n, v in zip(names, vlist)])
                                                          for vlist in vals]})
        # Fit models and plot signal
        # Remove True anchors, as they are non-existent
        try:
            models_val = test(Ksum=Ksum, Klist=Klist,
                          inxs=range(rank),
                          X=x, Xp=xv, y=y, f=f,  delta=delta, lbd=lbd,
                          methods=methods)
            models = test(Ksum=Ksum, Klist=Klist,
                          inxs=range(rank),
                          X=x, Xp=xp, y=y, f=f, delta=delta, lbd=lbd,
                          methods=methods)
        except Exception as e:
            print("Exception: %s %s" % (e, e.message))
            continue
        del models["True"]

        # Store file as pdf + eps
        fname = os.path.join(subdname, "plot_multi-%s_tsi-%d_lbd-%.3f_rank-%d.pdf" % (dataset, tsi, lbd, rank))
        plot_signal_subplots(X=xp, Xp=xp, y=yp, f=None, models=models, f_out=fname)

        fname = os.path.join(subdname, "plot_multi-%s_tsi-%d_lbd-%.3f_rank-%d.eps" % (dataset, tsi, lbd, rank))
        plot_signal_subplots(X=xp, Xp=xp, y=yp, f=None, models=models, f_out=fname)

        for ky in models.keys():
            mse_yv = mse(models_val[ky]["yp"].ravel(), yv.ravel())
            mse_yp = mse(models[ky]["yp"].ravel(), yp.ravel())

            time = models[ky]["time"]
            row = {"experiment": kernel,
                   "signal": dataset, "tsi": tsi, "method": ky,
                   "mse_f": 0, "mse_y": mse_yp, "mse_val": mse_yv,
                   "rank": rank, "lbd": lbd, "n": n, "time": time}
            writer.writerow(row)


if __name__ == "__main__":
    # Input arguments
    parser = argparse.ArgumentParser(description=hlp)
    parser.add_argument("dataset", help="Time series. One of {T1-T9}.")
    parser.add_argument("kernel",  help="Kernel function used for modelling. One of {exponential, periodic, matern}.")
    parser.add_argument("output",  help="Output directory.")
    args = parser.parse_args()

    # Output directory
    data_set = args.dataset
    kern = args.kernel
    out_dir = args.output
    process(data_set, kern, out_dir)

File: examples/energy/energy.R
hlp = "Post-processing of experiments with 1D time series (Energy dataset)."

require(optparse)
require(ggplot2)
require(scmamp)
require(xtable)

# Defined order of methods
row.order = c("Mklaren", "CSI", "ICD", "Nystrom", "RFF", "RFF-NS", "SPGP", "Arima")

# Parse input arguments
option_list = list(
  make_option(c("-i", "--input"), type="character", help="Results file (.csv)"),
  make_option(c("-o", "--output"), type="character", help="Output directory.")
);
opt_parser = OptionParser(option_list=option_list, description=hlp);
opt = parse_args(opt_parser);
in_file = opt$input
out_dir = opt$output
dir.create(out_dir, showWarnings = FALSE)

# Selection of methods
alldata = read.csv(in_file, header = TRUE, stringsAsFactors = FALSE) 
methods = unique(alldata$method)


# Select scores via cross-validation
alldata$name = sprintf("%s.%s.%s.%s", alldata$method, alldata$tsi, 
                                      alldata$signal, alldata$rank)
# Select rows via CV
af = aggregate(alldata$mse_val, by=list(method=alldata$method, 
                                tsi=alldata$tsi, 
                                signal=alldata$signal,
                                rank=alldata$rank), min)
row.names(af) <- sprintf("%s.%s.%s.%s", af$method, af$tsi, af$signal, af$rank)
alldata$best = af[alldata$name, "x"] == alldata$mse_val
data = alldata[alldata$best,]

# Validation RMSE per each rank
for (r in unique(data$rank)){
  k = unique(data$experiment)
  fname = file.path(out_dir, sprintf("test_kernel-%s_rank-%02d.pdf", k, r))
  dr = data[data$rank == r,]
  p = qplot(data=dr, x=as.factor(signal), 
        y=sqrt(mse_y), fill=method, geom="boxplot",
        xlab="Time series", ylab="RMSE", 
        ylim=c(0, 10), 
        main=sprintf("Rank=%d kernel=%s", r, k))
  ggsave(fname, width = 10, height = 4)
  message(sprintf("Written %s", fname))
}

# Store results in tables
for (r in unique(data$rank)){
  k = unique(data$experiment)
  dr = data[data$rank == r,]

  # Convert to RMSE
  agg.m = aggregate(sqrt(dr$mse_y), by=list(method=dr$method, dataset=dr$signal), mean)
  agg.s = aggregate(sqrt(dr$mse_y), by=list(method=dr$method, dataset=dr$signal), sd)
  
  signals = sort(unique(dr$signal))
  M = matrix(Inf, ncol=length(signals), nrow=length(methods))
  S = matrix(Inf, ncol=length(signals), nrow=length(methods))
  row.names(M) = methods
  colnames(M) = signals
  row.names(S) = methods
  colnames(S) = signals
  for
  (i in 1:nrow(agg.m)) M[agg.m[i,"method"], agg.m[i,"dataset"]] = agg.m[i, "x"]
  for (i in 1:nrow(agg.s)) S[agg.s[i,"method"], agg.s[i,"dataset"]] = agg.s[i, "x"]
  
  # Ranks 
  fname = file.path(out_dir, sprintf("cd_kernel-%s_rank-%02d.pdf", k, r))
  pdf(fname, width=8, height=5)
  plotCD(t(-M), alpha=.05)
  message(sprintf("Written %s", fname))
  
  # Friedman test on ranks
  message("Friedman test on ranks:")
  print(friedman.test(-M))
  
  # Text matrix 
  R = matrix("-", ncol=length(signals), nrow=length(row.order))
  row.names(R) = row.order
  colnames(R) = signals
  for(j in 1:ncol(M)){
    R[names(M[,j]), colnames(M)[j]] = sprintf("%.2f$\\pm$%.2f", M[,j], S[,j])
    vals = M[methods, j]
    best = names(which(vals == min(vals)))
    R[best, j] = sprintf("\\textbf{%.2f$\\pm$%.2f}", M[best,j], S[best,j])
  }
  
  # Store table
  fname = file.path(out_dir, sprintf("table_kernel-%s_rank-%02d.tex", k, r))
  tab = xtable(R)
  sink(fname)
  print(xtable(t(R)),
        sanitize.colnames.function=identity,
        sanitize.text.function=identity)
  sink()
  message(sprintf("Written %s", fname))
}